src/main/java/net/replaceitem/integratedcircuit/client/gui/IntegratedCircuitScreen.java

```diff
@@ -337,35 +432,42 @@ public boolean mouseDragged(double mouseX, double mouseY, int button, double del
         return super.mouseDragged(mouseX, mouseY, button, deltaX, deltaY);
     }
 
-
     private void rotateComponent(int amount) {
-        if(this.cursorState != null && this.cursorState.getComponent() instanceof FacingComponent) {
+        if (this.cursorState != null && this.cursorState.getComponent() instanceof FacingComponent) {
             this.cursorRotation = this.cursorRotation.rotated(amount);
             this.cursorState = this.cursorState.with(FacingComponent.FACING, this.cursorRotation);
         }
     }
 
     @Override
     public boolean mouseScrolled(double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
-        if(DefaultConfig.config.getInvertScrollDirection()) verticalAmount = -verticalAmount;
+        if (DefaultConfig.config.getInvertScrollDirection())
+            verticalAmount = -verticalAmount;
         int intAmount = (int) verticalAmount;
         switch (DefaultConfig.config.getScrollBehaviour()) {
             case ROTATE -> rotateComponent(-intAmount);
-            case SELECT_COMPONENT -> selectPalette(selectedComponentSlot - intAmount);
+            case SELECT_COMPONENT ->
+                selectPalette(toolbox.getSelectedToolSlot() - intAmount);
         }
         return true;
     }
-    
+
     @Override
     public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
-        if(matchesKey(DefaultConfig.config.getRotateKeybind(), keyCode, scanCode)) {
+        if (this.customNameTextField.isFocused()) {
```

Comment:
You can override the setFocused method of TextFieldWidget when constructing it and use that for detecting when it loses focus. Then you wouldn't need the extra checks in keyPressed or mouseClicked